Applications/Spire/Source/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,121 @@
+#include "Spire/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.hpp"
+#include "Spire/Spire/QtPromise.hpp"
+
+using namespace boost::icl;
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+CachedOrderImbalanceIndicatorModel::CachedOrderImbalanceIndicatorModel(
+  std::shared_ptr<OrderImbalanceIndicatorModel> m_source_model)
+  : m_source_model(std::move(m_source_model)) {}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::subscribe(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  if(contains(m_ranges, continuous_interval<ptime>::closed(start, end))) {
+    return get_subscription(start, end, slot);
+  }
+  return load_imbalances(start, end, slot);
+}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::get_subscription(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  m_subscriptions.push_back(Subscription(start, end));
+  auto first = std::find_if(m_imbalances.begin(), m_imbalances.end(),
+    [=] (const auto& imbalance) { return start <= imbalance.m_timestamp; });
+  auto last = std::find_if(m_imbalances.begin(), m_imbalances.end(),
+    [=] (const auto& imbalance) { return end < imbalance.m_timestamp; });
+  return {m_subscriptions.back().m_imbalance_signal.connect(slot),
+    QtPromise([=] {
+      return std::vector<OrderImbalance>(first, last);
+    })};
+}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::load_imbalances(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  auto promises = std::vector<QtPromise<std::vector<OrderImbalance>>>();
+  auto ranges = interval_set<ptime>(
+    {continuous_interval<ptime>::closed(start, end)}) - m_ranges;
```

Comment:
What do the curly brackets do in ```{continuous_interval...}```? What happens if you remove the curly brackets?

Reply:
I didn't realize there was a constructor for single values.